crates/typst/src/geom/color.rs

```diff
@@ -334,183 +997,687 @@ cast! {
 }
 
 /// A weight for color mixing.
-struct Weight(f32);
+struct Weight(f64);
 
 cast! {
     Weight,
-    v: f64 => Self(v as f32),
-    v: Ratio => Self(v.get() as f32),
+    v: f64 => Self(v),
+    v: Ratio => Self(v.get()),
 }
 
-/// Convert an RGBA color to four components in the given color space.
-fn rgba_to_vec4(color: RgbaColor, space: ColorSpace) -> [f32; 4] {
-    match space {
-        ColorSpace::Oklab => {
-            let RgbaColor { r, g, b, a } = color;
-            let oklab = oklab::srgb_to_oklab(oklab::RGB { r, g, b });
-            [oklab.l, oklab.a, oklab.b, a as f32 / 255.0]
+/// A color space for mixing.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum ColorSpace {
+    /// A perceptual color space.
+    Oklab,
+
+    /// The standard RGB color space.
+    Srgb,
+
+    /// The D65-gray color space.
+    D65Gray,
+
+    /// The linear RGB color space.
+    LinearRgb,
+
+    /// The HSL color space.
+    Hsl,
+
+    /// The HSV color space.
+    Hsv,
+
+    /// The CMYK color space.
+    Cmyk,
+}
+
+cast! {
+    ColorSpace,
+    self => match self {
+        Self::Oklab => Color::oklab_data(),
+        Self::Srgb => Color::rgb_data(),
+        Self::D65Gray => Color::luma_data(),
+        Self::LinearRgb => Color::linear_rgb_data(),
+        Self::Hsl => Color::hsl_data(),
+        Self::Hsv => Color::hsv_data(),
+        Self::Cmyk => Color::cmyk_data(),
+    }.into_value(),
+    v: Value => {
+        if !matches!(v, Value::Func(_)) {
+            bail!(
+                "expected `rgb`, `luma`, `cmyk`, `oklab`, `color.linear-rgb`, `color.hsl`, or `color.hsv`, found {}",
+                v.ty()
+            );
         }
-        ColorSpace::Srgb => {
-            let RgbaColor { r, g, b, a } = color;
-            [r as f32 / 255.0, g as f32 / 255.0, b as f32 / 255.0, a as f32 / 255.0]
+
+        let func = v
+            .cast::<Func>()
+            .unwrap()
+            .native()
+            .ok_or(
+                "expected `rgb`, `luma`, `cmyk`, `oklab`, `color.linear-rgb`, `color.hsl`, or `color.hsv`"
+            )?
+            .function;
+
+        // Here comparing the function pointer since it's `Eq`
+        // whereas the `NativeFuncData` is not.
+        if func == Color::oklab_data().function {
```

Comment:
We could implement `PartialEq<&'static NativeFuncData> for Func`. Then we can remove the `native` method, the casting code, and the duplicate error message and have this simpler.